File: src/commands/restore.ts
import { basename, dirname, isAbsolute, join, resolve } from "jsr:@std/path@1.1.4";
import { loadConfig } from "../config.ts";
import { CliError } from "../errors.ts";
import { appendLog } from "../log.ts";
import { ensureIntegrity } from "../sqlite.ts";
import type { RuntimeOptions } from "../types.ts";

export async function runRestore(
  archiveArg: string,
  options: RuntimeOptions = {},
): Promise<{ exitCode: number; stdout: string[]; stderr: string[] }> {
  const stdout: string[] = [];
  const stderr: string[] = [];

  try {
    const config = await loadConfig({ ...options, required: true });
    if (!config) {
      throw new CliError("config file not found", "ERR_CONFIG_NOT_FOUND", 1);
    }

    const browserRunning = isBrowserRunning(options);
    if (browserRunning) {
      throw new CliError(
        "Zen browser must be closed before restoring",
        "ERR_BROWSER_RUNNING",
        1,
      );
    }

    const archivePath = await resolveArchivePath(archiveArg, config.backup.local_path, options);
    const stagingDir = await extractArchiveSafely(archivePath);

    const now = options.now ?? new Date();
    const preRestorePath = await rotateProfileToPreRestore(config.profile.path, now);
    await Deno.mkdir(config.profile.path, { recursive: true });
    await copyDirContents(stagingDir, config.profile.path);
    await Deno.remove(stagingDir, { recursive: true }).catch(() => undefined);
    await validateSqliteFiles(config.profile.path);

    await appendLog(
      config.backup.local_path,
      "RESTORE",
      `restored profile from ${basename(archivePath)}`,
    );
    stdout.push(`Restored from archive: ${archivePath}`);
    stdout.push(`Pre-restore backup: ${preRestorePath}`);
    return { exitCode: 0, stdout, stderr };
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    stderr.push(message);
    return { exitCode: 1, stdout, stderr };
  }
}

async function resolveArchivePath(
  input: string,
  backupRoot: string,
  options: RuntimeOptions,
): Promise<string> {
  const cwd = options.cwd ?? Deno.cwd();
  const candidates = [
    isAbsolute(input) ? input : resolve(cwd, input),
    join(backupRoot, input),
    join(backupRoot, "daily", input),
    join(backupRoot, "weekly", input),
  ];

  for (const candidate of candidates) {
    if (await exists(candidate)) return candidate;
  }

  throw new CliError(`archive not found: ${input}`, "ERR_ARCHIVE_NOT_FOUND", 1);
}

async function extractArchiveSafely(archivePath: string): Promise<string> {
  const listed = await new Deno.Command("tar", {
    args: ["-tzf", archivePath],
    stdout: "piped",
    stderr: "piped",
  }).output();
  if (!listed.success) {
    throw new CliError(
      `invalid or corrupted archive: ${basename(archivePath)}`,
      "ERR_ARCHIVE_INVALID",
      1,
    );
  }

  const listing = new TextDecoder().decode(listed.stdout).trim();
  const entries = listing.length === 0 ? [] : listing.split("\n");
  for (const raw of entries) {
    const candidate = sanitizeTarEntry(raw);
    if (!candidate) continue;
    if (candidate.split("/").includes("..")) {
      throw new CliError(`invalid archive entry: ${raw}`, "ERR_ARCHIVE_INVALID", 1);
    }
    if (/^[A-Za-z]:\//.test(candidate)) {
      throw new CliError(`invalid archive entry: ${raw}`, "ERR_ARCHIVE_INVALID", 1);
    }
  }

  const stagingDir = await Deno.makeTempDir({ prefix: "zen-restore-staging-" });
  const extracted = await new Deno.Command("tar", {
    args: ["-xzf", archivePath, "-C", stagingDir],
    stdout: "null",
    stderr: "piped",
  }).output();
  if (!extracted.success) {
    await Deno.remove(stagingDir, { recursive: true }).catch(() => undefined);
    throw new CliError(
      `invalid or corrupted archive: ${basename(archivePath)}`,
      "ERR_ARCHIVE_INVALID",
      1,
    );
  }

  return stagingDir;
}

function sanitizeTarEntry(entry: string): string {
  let value = entry.trim().replaceAll("\\", "/");
  while (value.startsWith("./")) value = value.slice(2);
  while (value.startsWith("/")) value = value.slice(1);
  return value;
}

async function rotateProfileToPreRestore(profilePath: string, now: Date): Promise<string> {
  const stamp = now.toISOString().slice(0, 10);
  let candidate = `${profilePath}.pre-restore-${stamp}`;
  let index = 2;
  while (await exists(candidate)) {
    candidate = `${profilePath}.pre-restore-${stamp}-${index}`;
    index += 1;
  }

  if (await exists(profilePath)) {
    await Deno.rename(profilePath, candidate);
  } else {
    await Deno.mkdir(candidate, { recursive: true });
  }
  return candidate;
}

async function copyDirContents(fromDir: string, toDir: string): Promise<void> {
  for await (const entry of Deno.readDir(fromDir)) {
    const source = join(fromDir, entry.name);
    const target = join(toDir, entry.name);
    if (entry.isDirectory) {
      await Deno.mkdir(target, { recursive: true });
      await copyDirContents(source, target);
      continue;
    }
    if (entry.isFile) {
      await Deno.mkdir(dirname(target), { recursive: true });
      await Deno.copyFile(source, target);
    }
  }
}

async function validateSqliteFiles(root: string): Promise<void> {
  for await (const file of walkFiles(root)) {
    if (file.endsWith(".sqlite")) {
      await ensureIntegrity(file);
    }
  }
}

async function* walkFiles(root: string): AsyncGenerator<string> {
  for await (const entry of Deno.readDir(root)) {
    const path = join(root, entry.name);
    if (entry.isDirectory) {
      yield* walkFiles(path);
    } else if (entry.isFile) {
      yield path;
    }
  }
}

function isBrowserRunning(options: RuntimeOptions): boolean {
  return options.env?.ZEN_BACKUP_BROWSER_RUNNING === "1";
}

async function exists(path: string): Promise<boolean> {
  try {
    await Deno.stat(path);
    return true;
  } catch {
    return false;
  }
}

File: rust/zen-backup/src/commands/restore.rs
use crate::config::{load_config, ConfigError};
use chrono::{DateTime, Utc};
use std::ffi::OsStr;
use std::fs;
use std::fs::OpenOptions;
use std::io::Write;
use std::path::{Path, PathBuf};
use std::process::Command;

pub struct CommandOutput {
    pub exit_code: i32,
    pub stdout: String,
    pub stderr: String,
}

pub fn run_restore(archive_arg: &str, cwd: &Path) -> CommandOutput {
    let config = match load_config(true, cwd) {
        Ok(Some(v)) => v,
        Ok(None) => {
            return CommandOutput {
                exit_code: 1,
                stdout: String::new(),
                stderr: "config file not found".to_string(),
            }
        }
        Err(ConfigError::Parse(message)) | Err(ConfigError::NotFound(message)) => {
            return CommandOutput {
                exit_code: 1,
                stdout: String::new(),
                stderr: message,
            }
        }
    };

    if std::env::var("ZEN_BACKUP_BROWSER_RUNNING").ok().as_deref() == Some("1") {
        return CommandOutput {
            exit_code: 1,
            stdout: String::new(),
            stderr: "Zen browser must be closed before restoring".to_string(),
        };
    }

    let archive_path =
        match resolve_archive_path(archive_arg, cwd, Path::new(&config.backup_local_path)) {
            Some(v) => v,
            None => {
                return CommandOutput {
                    exit_code: 1,
                    stdout: String::new(),
                    stderr: format!("archive not found: {archive_arg}"),
                }
            }
        };

    let staging_dir = match extract_archive_safely(&archive_path) {
        Ok(v) => v,
        Err(err) => {
            return CommandOutput {
                exit_code: 1,
                stdout: String::new(),
                stderr: err,
            }
        }
    };

    let profile_path = PathBuf::from(&config.profile_path);
    let pre_restore = match rotate_profile_to_prerestore(&profile_path) {
        Ok(v) => v,
        Err(err) => {
            return CommandOutput {
                exit_code: 1,
                stdout: String::new(),
                stderr: err,
            }
        }
    };

    if fs::create_dir_all(&profile_path).is_err() {
        return CommandOutput {
            exit_code: 1,
            stdout: String::new(),
            stderr: "failed to create profile directory".to_string(),
        };
    }

    if copy_dir_contents(&staging_dir, &profile_path).is_err() {
        return CommandOutput {
            exit_code: 1,
            stdout: String::new(),
            stderr: "failed to restore archive contents".to_string(),
        };
    }
    let _ = fs::remove_dir_all(&staging_dir);

    if validate_sqlite_files(&profile_path).is_err() {
        return CommandOutput {
            exit_code: 1,
            stdout: String::new(),
            stderr: format!(
                "invalid or corrupted archive: {}",
                archive_path
                    .file_name()
                    .unwrap_or_else(|| OsStr::new("archive"))
                    .to_string_lossy()
            ),
        };
    }

    append_restore_log(
        Path::new(&config.backup_local_path),
        &archive_path
            .file_name()
            .unwrap_or_else(|| OsStr::new("archive"))
            .to_string_lossy(),
    );

    CommandOutput {
        exit_code: 0,
        stdout: format!(
            "Restored from archive: {}\nPre-restore backup: {}",
            archive_path.display(),
            pre_restore.display()
        ),
        stderr: String::new(),
    }
}

fn append_restore_log(backup_root: &Path, archive_name: &str) {
    let log_path = backup_root.join("backup.log");
    let _ = fs::create_dir_all(backup_root);
    let mut file = match OpenOptions::new().create(true).append(true).open(log_path) {
        Ok(file) => file,
        Err(_) => return,
    };
    let _ = writeln!(
        file,
        "[{}] RESTORE: restored {}",
        now_iso_utc(),
        archive_name
    );
}

fn now_iso_utc() -> String {
    if let Ok(raw) = std::env::var("ZEN_BACKUP_TEST_NOW") {
        if let Ok(value) = DateTime::parse_from_rfc3339(&raw) {
            return value.with_timezone(&Utc).to_rfc3339();
        }
    }
    Utc::now().to_rfc3339()
}

fn resolve_archive_path(input: &str, cwd: &Path, backup_root: &Path) -> Option<PathBuf> {
    let raw = PathBuf::from(input);
    let candidates = vec![
        if raw.is_absolute() {
            raw.clone()
        } else {
            cwd.join(&raw)
        },
        backup_root.join(input),
        backup_root.join("daily").join(input),
        backup_root.join("weekly").join(input),
    ];
    candidates.into_iter().find(|p| p.exists())
}

fn extract_archive_safely(archive: &Path) -> Result<PathBuf, String> {
    let list = Command::new("tar")
        .arg("-tzf")
        .arg(archive)
        .output()
        .map_err(|_| "invalid or corrupted archive".to_string())?;
    if !list.status.success() {
        return Err(format!(
            "invalid or corrupted archive: {}",
            archive
                .file_name()
                .unwrap_or_else(|| OsStr::new("archive"))
                .to_string_lossy()
        ));
    }
    let listing = String::from_utf8_lossy(&list.stdout);
    for line in listing.lines() {
        let entry = sanitize_tar_entry(line);
        if entry.split('/').any(|s| s == "..") || is_windows_abs(&entry) {
            return Err(format!("invalid archive entry: {line}"));
        }
    }
    let staging = unique_temp_dir("zen-restore-staging");
    let _ = fs::create_dir_all(&staging);
    let extract = Command::new("tar")
        .arg("-xzf")
        .arg(archive)
        .arg("-C")
        .arg(&staging)
        .output()
        .map_err(|_| "invalid or corrupted archive".to_string())?;
    if !extract.status.success() {
        let _ = fs::remove_dir_all(&staging);
        return Err(format!(
            "invalid or corrupted archive: {}",
            archive
                .file_name()
                .unwrap_or_else(|| OsStr::new("archive"))
                .to_string_lossy()
        ));
    }
    Ok(staging)
}

fn sanitize_tar_entry(entry: &str) -> String {
    let mut value = entry.trim().replace('\\', "/");
    while value.starts_with("./") {
        value = value[2..].to_string();
    }
    while value.starts_with('/') {
        value = value[1..].to_string();
    }
    value
}

fn is_windows_abs(path: &str) -> bool {
    path.len() > 3 && path.as_bytes()[1] == b':' && path.as_bytes()[2] == b'/'
}

fn rotate_profile_to_prerestore(profile: &Path) -> Result<PathBuf, String> {
    let date = std::env::var("ZEN_BACKUP_TEST_NOW")
        .ok()
        .map(|v| v.chars().take(10).collect::<String>())
        .unwrap_or_else(|| "1970-01-01".to_string());
    let mut candidate = PathBuf::from(format!("{}.pre-restore-{}", profile.display(), date));
    let mut idx = 2;
    while candidate.exists() {
        candidate = PathBuf::from(format!(
            "{}.pre-restore-{}-{}",
            profile.display(),
            date,
            idx
        ));
        idx += 1;
    }
    if profile.exists() {
        fs::rename(profile, &candidate).map_err(|_| "failed to rotate profile".to_string())?;
    } else {
        fs::create_dir_all(&candidate)
            .map_err(|_| "failed to create pre-restore dir".to_string())?;
    }
    Ok(candidate)
}

fn copy_dir_contents(from: &Path, to: &Path) -> Result<(), String> {
    for entry in fs::read_dir(from).map_err(|_| "failed to read staging dir".to_string())? {
        let entry = entry.map_err(|_| "failed to read staging entry".to_string())?;
        let src = entry.path();
        let dst = to.join(entry.file_name());
        if src.is_dir() {
            fs::create_dir_all(&dst).map_err(|_| "failed to create restore dir".to_string())?;
            copy_dir_contents(&src, &dst)?;
        } else if src.is_file() {
            if let Some(parent) = dst.parent() {
                let _ = fs::create_dir_all(parent);
            }
            fs::copy(&src, &dst).map_err(|_| "failed to copy restore file".to_string())?;
        }
    }
    Ok(())
}

fn validate_sqlite_files(root: &Path) -> Result<(), ()> {
    for file in walk_files(root).map_err(|_| ())? {
        if file.to_string_lossy().ends_with(".sqlite") {
            let out = Command::new("sqlite3")
                .arg(&file)
                .arg("PRAGMA integrity_check;")
                .output()
                .map_err(|_| ())?;
            if !out.status.success() {
                return Err(());
            }
            let stdout = String::from_utf8_lossy(&out.stdout).to_lowercase();
            if !stdout.contains("ok") {
                return Err(());
            }
        }
    }
    Ok(())
}

fn walk_files(root: &Path) -> Result<Vec<PathBuf>, std::io::Error> {
    let mut out = vec![];
    for entry in fs::read_dir(root)? {
        let path = entry?.path();
        if path.is_dir() {
            out.extend(walk_files(&path)?);
        } else if path.is_file() {
            out.push(path);
        }
    }
    Ok(out)
}

fn unique_temp_dir(prefix: &str) -> PathBuf {
    use std::time::{SystemTime, UNIX_EPOCH};
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|v| v.as_nanos())
        .unwrap_or(0);
    std::env::temp_dir().join(format!("{prefix}-{nanos}"))
}
